File: examples/php/Case08.php
<?php
declare(strict_types=1);

final class Case08Record
{
    public function __construct(public string $state, public int $revision) {}
}

final class Case08Store
{
    public function __construct(public Case08Record $record) {}
}

final class Case08WeakProjection
{
    private int $guard = 0; // Lost with the process.
    public function __construct(private Case08Store $store) {}

    // faultscope:begin fs-c08.memory-only-revision
    public function apply(Case08Record $event): bool
    {
        if ($event->revision <= $this->guard) return false;
        $this->store->record = new Case08Record($event->state, $this->store->record->revision);
        $this->guard = $event->revision; // Revision is not durable.
        return true;
    }
    // faultscope:end fs-c08.memory-only-revision
}

// faultscope:begin fs-c08.persist-state-with-revision
function case08PersistStateWithRevision(Case08Store $store, Case08Record $next): void
{
    $store->record = $next; // One atomic durable record in this model.
}
// faultscope:end fs-c08.persist-state-with-revision

final class Case08StrongProjection
{
    private bool $ready = false;
    private Case08Record $current;
    private function __construct(private Case08Store $store) {}

    // faultscope:begin fs-c08.recover-state-with-revision
    public static function recover(Case08Store $store): self
    {
        $projection = new self($store);
        $projection->current = $store->record; // Restore value and revision together.
        $projection->ready = true; // Only after recovery may events be applied.
        return $projection;
    }
    // faultscope:end fs-c08.recover-state-with-revision

    // faultscope:begin fs-c08.reject-stale-after-restart
    public function apply(Case08Record $event): bool
    {
        if (!$this->ready || $event->revision <= $this->current->revision) return false;
        case08PersistStateWithRevision($this->store, $event);
        $this->current = $event;
        return true;
    }
    // faultscope:end fs-c08.reject-stale-after-restart
}

// faultscope:begin fs-c08.rebuild-before-ready
function case08RebuildBeforeReady(array $history): Case08StrongProjection
{
    $store = new Case08Store(new Case08Record('', 0));
    foreach ($history as $event) { // Complete authoritative history; no external effects here.
        if ($event->revision > $store->record->revision) case08PersistStateWithRevision($store, $event);
    }
    return Case08StrongProjection::recover($store); // Ready only after replay.
}
// faultscope:end fs-c08.rebuild-before-ready

File: examples/c/case08.c
#include <string.h>
#include <threads.h>

typedef struct {
    char state[24];
    int revision;
} Case08Record;

typedef struct {
    Case08Record record;
    mtx_t mutex;
} Case08Store;

static void case08_store_init(Case08Store *store) {
    store->record = (Case08Record){{0}, 0};
    mtx_init(&store->mutex, mtx_plain);
}

typedef struct {
    Case08Store *store;
    int guard; /* Lost with the process. */
} Case08WeakProjection;

// faultscope:begin fs-c08.memory-only-revision
int case08_memory_only_revision(Case08WeakProjection *projection, Case08Record event) {
    if (event.revision <= projection->guard) return 0;
    mtx_lock(&projection->store->mutex);
    memcpy(projection->store->record.state, event.state, sizeof(event.state)); /* Value only. */
    mtx_unlock(&projection->store->mutex);
    projection->guard = event.revision;
    return 1;
}
// faultscope:end fs-c08.memory-only-revision

// faultscope:begin fs-c08.persist-state-with-revision
void case08_persist_state_with_revision(Case08Store *store, Case08Record next) {
    mtx_lock(&store->mutex);
    store->record = next; /* One coherent durable record in this model. */
    mtx_unlock(&store->mutex);
}
// faultscope:end fs-c08.persist-state-with-revision

typedef struct {
    Case08Store *store;
    Case08Record current;
    int ready;
} Case08StrongProjection;

// faultscope:begin fs-c08.recover-state-with-revision
Case08StrongProjection case08_recover_state_with_revision(Case08Store *store) {
    Case08StrongProjection projection = {.store = store, .ready = 0};
    mtx_lock(&store->mutex);
    projection.current = store->record; /* Restore value and revision together. */
    mtx_unlock(&store->mutex);
    projection.ready = 1; /* Only after recovery may events be applied. */
    return projection;
}
// faultscope:end fs-c08.recover-state-with-revision

// faultscope:begin fs-c08.reject-stale-after-restart
int case08_reject_stale_after_restart(Case08StrongProjection *projection, Case08Record event) {
    if (!projection->ready || event.revision <= projection->current.revision) return 0;
    case08_persist_state_with_revision(projection->store, event);
    projection->current = event;
    return 1;
}
// faultscope:end fs-c08.reject-stale-after-restart

// faultscope:begin fs-c08.rebuild-before-ready
Case08StrongProjection case08_rebuild_before_ready(Case08Store *store,
                                                    const Case08Record *history, int count) {
    case08_persist_state_with_revision(store, (Case08Record){{0}, 0});
    for (int index = 0; index < count; index++) { /* Complete authoritative history. */
        if (history[index].revision > store->record.revision)
            case08_persist_state_with_revision(store, history[index]);
    }
    return case08_recover_state_with_revision(store); /* Ready only after replay. */
}
// faultscope:end fs-c08.rebuild-before-ready
